### PR TITLE
Fix/mac not found in db 13.x

### PR DESCRIPTION
**Pull request by fdurand**
# Description
Update ip4log directly from pfdhcp when the network is managed by PacketFence.

# Impacts
pfqueue + pfdhcp

# Delete branch after merge
YES

# Checklist
- [ ] Document the feature
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries

## Enhancements
* Update ip4log from pfdhcp when the network is managed by PacketFence

### CODE REVIEW

File: go/cmd/pfdhcp/utils.go [hunk 1]
```diff
@@ -430,6 +432,70 @@ func IsIPv6(address net.IP) bool {
 	return strings.Count(address.String(), ":") >= 2
 }
 
+// MysqlUpdateIP4Log update the ip4log table
+func MysqlUpdateIP4Log(mac string, ip string, duration time.Duration) error {
+	if err := MySQLdatabase.PingContext(ctx); err != nil {
+		log.LoggerWContext(ctx).Error("Unable to ping database, reconnect: " + err.Error())
+	}
+
+	MAC2IP, err := MySQLdatabase.Prepare("SELECT ip FROM ip4log WHERE mac = ? AND (end_time = " + ZeroDate + " OR ( end_time + INTERVAL 30 SECOND ) > NOW()) ORDER BY start_time DESC LIMIT 1")
```
Comment (jrouzierinverse) [inline go/cmd/pfdhcp/utils.go]: Quote the Zero Date
```go
	MAC2IP, err := MySQLdatabase.Prepare("SELECT ip FROM ip4log WHERE mac = ? AND (end_time = '" + ZeroDate + "' OR ( end_time + INTERVAL 30 SECOND ) > NOW()) ORDER BY start_time DESC LIMIT 1")

 ```

File: go/cmd/pfdhcp/utils.go [hunk 2]
```diff
@@ -430,6 +432,70 @@ func IsIPv6(address net.IP) bool {
...
+	if err != nil {
+		return err
+	}
+
+	IP2MAC, err := MySQLdatabase.Prepare("SELECT mac FROM ip4log WHERE ip = ? AND (end_time = " + ZeroDate + " OR end_time > NOW()) ORDER BY start_time DESC")
```
Comment (jrouzierinverse) [inline go/cmd/pfdhcp/utils.go]: Quote the ZeroDate
```go
	IP2MAC, err := MySQLdatabase.Prepare("SELECT mac FROM ip4log WHERE ip = ? AND (end_time = '" + ZeroDate + "' OR end_time > NOW()) ORDER BY start_time DESC")
```

File: go/cmd/pfdhcp/utils.go [hunk 3]
```diff
@@ -45,7 +47,7 @@ func initiaLease(dhcpHandler *DHCPHandler, ConfNet pfconfigdriver.RessourseNetwo
 	binary.BigEndian.PutUint32(a, endip)
 	ipend := net.IPv4(a[0], a[1], a[2], a[3])
 
-	rows, err := MySQLdatabase.Query("select ip,mac,end_time,start_time from ip4log i where inet_aton(ip) between inet_aton(?) and inet_aton(?) and (end_time = 0 OR  end_time > NOW()) and end_time in (select MAX(end_time) from ip4log where mac = i.mac)  ORDER BY mac,end_time desc", dhcpHandler.start.String(), ipend.String())
+	rows, err := MySQLdatabase.Query("select ip,mac,end_time,start_time from ip4log i where inet_aton(ip) between inet_aton(?) and inet_aton(?) and (end_time = "+ZeroDate+" OR  end_time > NOW()) and end_time in (select MAX(end_time) from ip4log where mac = i.mac)  ORDER BY mac,end_time desc", dhcpHandler.start.String(), ipend.String())
```
Comment (jrouzierinverse) [inline go/cmd/pfdhcp/utils.go]: Quote the zero date
```go
	rows, err := MySQLdatabase.Query("select ip,mac,end_time,start_time from ip4log i where inet_aton(ip) between inet_aton(?) and inet_aton(?) and (end_time = '"+ZeroDate+"' OR  end_time > NOW()) and end_time in (select MAX(end_time) from ip4log where mac = i.mac)  ORDER BY mac,end_time desc", dhcpHandler.start.String(), ipend.String())
```

File: go/cmd/pfdhcp/utils.go [hunk 4]
```diff
@@ -430,6 +432,70 @@ func IsIPv6(address net.IP) bool {
...
+	if err != nil {
+		return err
+	}
+
+	IPClose, err := MySQLdatabase.Prepare(" UPDATE ip4log SET end_time = NOW() WHERE ip = ?")
+	if err != nil {
+		return err
+	}
+
+	IPInsert, err := MySQLdatabase.Prepare("INSERT INTO ip4log (mac, ip, start_time, end_time) VALUES (?, ?, NOW(), DATE_ADD(NOW(), INTERVAL ? SECOND)) ON DUPLICATE KEY UPDATE mac=VALUES(mac), end_time=VALUES(end_time)")
```
Comment (jrouzierinverse) [inline go/cmd/pfdhcp/utils.go]: Also, update the start_time

File: go/cmd/pfdhcp/utils.go [hunk 5]
```diff
@@ -430,6 +432,70 @@ func IsIPv6(address net.IP) bool {
...
+	if len(oldIP) > 0 && (oldIP != ip) {
+		_, err = IPClose.Exec(oldIP)
+		if err != nil {
+			return err
+		}
+	}
+	_, err = IPInsert.Exec(mac, ip, duration.Seconds())
+	if err != nil {
+		log.LoggerWContext(ctx).Info(err.Error())
+		_, err = IPUpdate.Exec(mac, duration.Seconds(), ip)
```
Comment (jrouzierinverse) [inline go/cmd/pfdhcp/utils.go]: Since IPInsert handle both insert and update remove IPUpdate

**Review by jrouzierinverse — CHANGES_REQUESTED** [review]

**Comment by fdurand** [comment]
BUMP

**Pull request merged**